examples-draft/SimpleExampleWF_Max.tla

```diff
@@ -0,0 +1,213 @@
+--------------------------- MODULE SimpleExampleWF_Max -----------------------
+(*
+That's an example of a proof of Spec => <>P. Such proof has to be based on
+a fairness condition. The general idea is that we have to show that either
+<>P is true, or an action is enabled that will make it true later.
+
+The example here defines an operator Max, then an algorithm that
+"calculates" the maximum in several steps (MaxAsFSM) and finally proves
+that the algorithm eventually will give the same result as the
+operator Max (a kind of refinement). In this way we model a non-atomic
+implementation of an operation defined in a declarative way.
+
+In the example SimpleExampleWF_Flag the variable was a boolean, thus we
+managed to prove the P \/ ENABLED A just from the definitions of P and A.
+Here we cannot do that and induction is needed to show P \/ ENABLED A as
+an inductive invariant.
+*)
+EXTENDS Naturals, TLAPS
+
+Max(a, b) ==
+  IF a > b THEN a ELSE b 
+
+MaxProps == \A a, b \in Nat:
+              /\ Max(a, b) >= a
+              /\ Max(a, b) >= b
+              /\ Max(a, b) \in Nat
+              /\ Max(a, b) \in {a, b}
+THEOREM MaxIsGEQ == MaxProps
+  BY DEF MaxProps, Max
+
+-----------------------------------------------------------------------------
+
+---------- MODULE MaxAsFSM ---------------
+EXTENDS Naturals
+CONSTANTS a, b
+ASSUME Assms == a \in Nat /\ b \in Nat
+VARIABLES x
+Null == CHOOSE n : n \notin Nat
+Init == x = Null
+Next ==
+ /\ x = Null
+ /\ \/ a >  b /\ x' = a
+    \/ a =< b /\ x' = b
+Live == WF_x(Next)
```

Comment:
A suggestion: instead of baking the proofs of the ENABLED condition into the proofs below, why not prove

LEMMA EnabledNext == x = Null => ENABLED <<Next>>_x

(you may also prove the equivalence, but this is the implication that you need).
In order to do that, show that

Next => << Next >>_x

then use ENABLEDrules to reduce the proof of the assertion to that of

x = Null => ENABLED Next

which (I believe) should be proved by ENABLEDrewrite.

Reply:
I'm sorry for sending the above comment to early. Applying `ExpandENABLED` and adding `mX = mX'` made the step passing.

Nevertheless it would be really useful if you could point to the relevant parts in the source code to understand ENABLEDrewrites and ENABLEDrules and related directives.